src/lib/ecore/efl_promise2.c

```diff
@@ -0,0 +1,277 @@
+static int FUTURE_EVENT_ID = -1;
+static Ecore_Event_Handler *_future_event_handler = NULL;
+
+static void
+_efl_future2_dispatch(Efl_Future2 *f, Eina_Value *value, Eina_Bool rejected)
+{
+   while (f)
+     {
+        if (((!rejected && f->success) || (rejected && f->error)))
+          break;
+        Efl_Future2 *n = f->next;
+        Eina_Free_Cb free_cb = f->free;
+        const void *data = f->data;
+        if (n) n->prev = NULL;
+        f->next = NULL;
+        eina_mempool_free(_future_mp, f);
+        if (free_cb) free((void *)data);
+        f = n;
+     }
+
+   if (!f)
+     {
+        eina_value_free(value);
+        return;
+     }
+
+   Eina_Value *next_value = rejected ? f->error((void *)f->data, value) :
+     f->success((void *)f->data, value);
+   if (value != next_value) eina_value_free(value);
+   if (f->free) f->free((void *)f->data);
+
+   Efl_Future2 *next_future = f->next;
+   if (next_future) next_future->prev = NULL;
+   eina_mempool_free(_future_mp, f);
+
+   if (!next_future)
+     {
+        eina_value_free(next_value);
+        return;
+     }
+
+   if (next_value && next_value->type == EINA_VALUE_TYPE_PROMISE2)
+     {
+        Efl_Promise2 *p = eina_value_promise2_steal(next_value);
+        eina_value_free(next_value);
+        f->reject = rejected;
+        p->future = next_future;
+     }
+   else if (((next_value && next_value->type == EINA_VALUE_TYPE_ERROR) || (rejected)))
+     _efl_future2_dispatch(next_future, next_value, EINA_TRUE);
+   else
+     _efl_future2_dispatch(next_future, next_value, EINA_FALSE);
+}
+
+static Eina_Bool
+_efl_future2_handler_cb(void *data EINA_UNUSED,
+                        int type EINA_UNUSED,
+                        void *event)
+{
+   Efl_Future2 *f = event;
+   _efl_future2_dispatch(f, f->pending_value, f->reject);
+   _pending_futures = eina_list_remove(_pending_futures, f);
+   return ECORE_CALLBACK_DONE;
+}
+
+static void
+_dummy_free(void *user_data EINA_UNUSED, void *func_data EINA_UNUSED)
+{
+}
+
+static Eina_Bool
+_efl_future2_schedule(Efl_Future2 *f, Eina_Value *value, Eina_Bool reject)
+{
+   EINA_SAFETY_ON_TRUE_RETURN_VAL(f->event != NULL, EINA_FALSE);
+
+   f->reject = reject;
+   if (value && value->type == EINA_VALUE_TYPE_PROMISE2)
+     {
+        Efl_Promise2 *p = eina_value_promise2_steal(value);
+        eina_value_free(value);
+        p->future = f;
+     }
+   else
+     {
+        f->event = ecore_event_add(FUTURE_EVENT_ID, f, _dummy_free, NULL);
+        EINA_SAFETY_ON_NULL_RETURN_VAL(f->event, EINA_FALSE);
+        f->pending_value = value;
+        _pending_futures = eina_list_append(_pending_futures, f);
+     }
+   return EINA_TRUE;
+}
+
+Eina_Bool
+efl_promise2_init(void)
+{
+   const char *choice;
+
+   choice = getenv("EINA_MEMPOOL");
+   if ((!choice) || (!choice[0]))
+     choice = "chained_mempool";
+
+   _promise_mp = eina_mempool_add(choice, "Efl_Promise2",
+                                  NULL, sizeof(Efl_Promise2), 16);
```

Comment:
I thought the same thing to be honest, however I saw other places (in ecore) using this size so I went with it. Let's increase it for now and leave a FIXME note.